File: minecraft-web-manager/backend/src/services/plugin.service.js
import fs from 'fs/promises';
import path from 'path';
import ConfigService from './config.service.js';

class PluginService {
  constructor() {
    this.serverPath = process.env.SERVER_PATH;
    this.pluginsPath = path.join(this.serverPath, 'plugins');
    this.configService = new ConfigService();
  }

  /**
   * Listar plugins instalados
   */
  async listPlugins() {
    try {
      const files = await fs.readdir(this.pluginsPath);
      const plugins = [];

      for (const file of files) {
        if (file.endsWith('.jar')) {
          const pluginPath = path.join(this.pluginsPath, file);
          const stats = await fs.stat(pluginPath);

          plugins.push({
            name: file.replace('.jar', ''),
            filename: file,
            enabled: true,
            size: Math.round(stats.size / 1024), // KB
            modified: stats.mtime
          });
        } else if (file.endsWith('.jar.disabled')) {
          const pluginPath = path.join(this.pluginsPath, file);
          const stats = await fs.stat(pluginPath);

          plugins.push({
            name: file.replace('.jar.disabled', ''),
            filename: file,
            enabled: false,
            size: Math.round(stats.size / 1024),
            modified: stats.mtime
          });
        }
      }

      return plugins;
    } catch (error) {
      throw new Error(`Error al listar plugins: ${error.message}`);
    }
  }

  /**
   * Toggle plugin enabled/disabled
   */
  async togglePlugin(pluginName) {
    try {
      const enabledPath = path.join(this.pluginsPath, `${pluginName}.jar`);
      const disabledPath = path.join(this.pluginsPath, `${pluginName}.jar.disabled`);

      try {
        await fs.access(enabledPath);
        // Está habilitado, deshabilitar
        await fs.rename(enabledPath, disabledPath);
        return { enabled: false, message: 'Plugin deshabilitado' };
      } catch {
        // Está deshabilitado, habilitar
        await fs.access(disabledPath);
        await fs.rename(disabledPath, enabledPath);
        return { enabled: true, message: 'Plugin habilitado' };
      }
    } catch (error) {
      throw new Error(`Error al cambiar estado del plugin: ${error.message}`);
    }
  }

  /**
   * Eliminar plugin
   */
  async deletePlugin(pluginName) {
    try {
      const jarPath = path.join(this.pluginsPath, `${pluginName}.jar`);
      const disabledPath = path.join(this.pluginsPath, `${pluginName}.jar.disabled`);

      try {
        await fs.unlink(jarPath);
      } catch {
        await fs.unlink(disabledPath);
      }

      // Eliminar carpeta de configuración si existe
      const configPath = path.join(this.pluginsPath, pluginName);
      try {
        await fs.rm(configPath, { recursive: true, force: true });
      } catch {
        // No existe, continuar
      }

      return { success: true, message: 'Plugin eliminado' };
    } catch (error) {
      throw new Error(`Error al eliminar plugin: ${error.message}`);
    }
  }

  /**
   * Obtener archivos de configuración de un plugin
   */
  async getPluginConfigFiles(pluginName) {
    try {
      const configPath = path.join(this.pluginsPath, pluginName);

      try {
        const files = await fs.readdir(configPath);
        const configFiles = files.filter(f => f.endsWith('.yml') || f.endsWith('.yaml') || f.endsWith('.json'));
        return configFiles;
      } catch {
        return [];
      }
    } catch (error) {
      throw new Error(`Error al obtener archivos de configuración: ${error.message}`);
    }
  }

  /**
   * Leer archivo de configuración de plugin
   */
  async readPluginConfig(pluginName, configFile) {
    try {
      const configPath = path.join(this.pluginsPath, pluginName, configFile);

      if (configFile.endsWith('.yml') || configFile.endsWith('.yaml')) {
        return await this.configService.readYaml(configPath);
      } else if (configFile.endsWith('.json')) {
        return await this.configService.readJson(configPath);
      } else {
        const content = await fs.readFile(configPath, 'utf-8');
        return content;
      }
    } catch (error) {
      throw new Error(`Error al leer configuración: ${error.message}`);
    }
  }

  /**
   * Escribir archivo de configuración de plugin
   */
  async writePluginConfig(pluginName, configFile, content) {
    try {
      const configPath = path.join(this.pluginsPath, pluginName, configFile);

      if (configFile.endsWith('.yml') || configFile.endsWith('.yaml')) {
        await this.configService.writeYaml(configPath, content);
      } else if (configFile.endsWith('.json')) {
        await this.configService.writeJson(configPath, content);
      } else {
        await fs.writeFile(configPath, content, 'utf-8');
      }

      return { success: true, message: 'Configuración actualizada' };
    } catch (error) {
      throw new Error(`Error al escribir configuración: ${error.message}`);
    }
  }
}

export default PluginService;

File: minecraft-web-manager/backend-python/app/services/plugin_service.py
"""Servicio para gestión de plugins"""
from pathlib import Path
from typing import Dict, Any, List
from app.core.config import settings


class PluginService:
    """Servicio para gestión de plugins"""
    
    def __init__(self):
        self.server_path = Path(settings.SERVER_PATH)
        self.plugins_path = self.server_path / "plugins"
    
    async def list_plugins(self) -> List[Dict[str, Any]]:
        """
        Listar plugins instalados
        
        Returns:
            Lista de plugins con nombre, enabled, size
        """
        plugins = []
        
        if not self.plugins_path.exists():
            return plugins
        
        for file in self.plugins_path.iterdir():
            if file.is_file() and (file.suffix == ".jar" or file.name.endswith(".jar.disabled")):
                enabled = file.suffix == ".jar"
                name = file.stem if enabled else file.name.replace(".jar.disabled", "")
                
                plugins.append({
                    "name": name,
                    "filename": file.name,
                    "enabled": enabled,
                    "size_mb": file.stat().st_size / (1024 * 1024),
                })
        
        return plugins
    
    async def toggle_plugin(self, plugin_name: str) -> Dict[str, Any]:
        """
        Enable/disable plugin (renombrar .jar <-> .jar.disabled)
        
        Args:
            plugin_name: Nombre del plugin
        
        Returns:
            Dict con success y nuevo estado
        """
        jar_file = self.plugins_path / f"{plugin_name}.jar"
        disabled_file = self.plugins_path / f"{plugin_name}.jar.disabled"
        
        try:
            if jar_file.exists():
                # Deshabilitar
                jar_file.rename(disabled_file)
                return {
                    "success": True,
                    "enabled": False,
                    "message": f"Plugin '{plugin_name}' deshabilitado"
                }
            elif disabled_file.exists():
                # Habilitar
                disabled_file.rename(jar_file)
                return {
                    "success": True,
                    "enabled": True,
                    "message": f"Plugin '{plugin_name}' habilitado"
                }
            else:
                return {"success": False, "message": "Plugin no encontrado"}
                
        except Exception as e:
            return {"success": False, "message": f"Error: {str(e)}"}
    
    async def upload_plugin(self, filename: str, content: bytes) -> Dict[str, Any]:
        """
        Subir nuevo plugin
        
        Args:
            filename: Nombre del archivo .jar
            content: Contenido del archivo
        
        Returns:
            Dict con success y mensaje
        """
        try:
            # Asegurar que la carpeta plugins existe
            self.plugins_path.mkdir(parents=True, exist_ok=True)
            
            # Guardar archivo
            plugin_file = self.plugins_path / filename
            plugin_file.write_bytes(content)
            
            return {
                "success": True,
                "message": f"Plugin '{filename}' subido exitosamente"
            }
            
        except Exception as e:
            return {"success": False, "message": f"Error al subir plugin: {str(e)}"}
    
    async def delete_plugin(self, plugin_name: str) -> Dict[str, Any]:
        """
        Eliminar plugin
        
        Args:
            plugin_name: Nombre del plugin
        
        Returns:
            Dict con success y mensaje
        """
        jar_file = self.plugins_path / f"{plugin_name}.jar"
        disabled_file = self.plugins_path / f"{plugin_name}.jar.disabled"
        
        try:
            if jar_file.exists():
                jar_file.unlink()
                return {
                    "success": True,
                    "message": f"Plugin '{plugin_name}' eliminado"
                }
            elif disabled_file.exists():
                disabled_file.unlink()
                return {
                    "success": True,
                    "message": f"Plugin '{plugin_name}' eliminado"
                }
            else:
                return {"success": False, "message": "Plugin no encontrado"}
                
        except Exception as e:
            return {"success": False, "message": f"Error: {str(e)}"}


# Instancia global
plugin_service = PluginService()
